File: App/Resources/AlgorithmDetails/mergesort/mergesort.js
function sort(array) {
  const scratch = new Array(array.length);

  function merge(start, mid, end) {
    let left = start;
    let right = mid;
    let out = start;
    while (left < mid && right < end) {
      if (array[left] <= array[right]) {
        scratch[out++] = array[left++];
      } else {
        scratch[out++] = array[right++];
      }
    }
    while (left < mid) scratch[out++] = array[left++];
    while (right < end) scratch[out++] = array[right++];
    for (let i = start; i < end; i++) array[i] = scratch[i];
  }

  function mergeSort(start, end) {
    if (end - start < 2) return;
    const mid = start + Math.floor((end - start) / 2);
    mergeSort(start, mid);
    mergeSort(mid, end);
    merge(start, mid, end);
  }

  mergeSort(0, array.length);
  return array;
}

var array = [0, 39, 21, 62, 91, 77, 14, 23, 90, 69, 51, 81, 68, 83, 32, 56];
sort(array);
console.log("[" + array.join(", ") + "]");

File: App/Resources/AlgorithmDetails/mergesort/mergesort.rb
def merge(array, scratch, start, mid, finish)
  left = start
  right = mid
  out = start
  while left < mid && right < finish
    if array[left] <= array[right]
      scratch[out] = array[left]
      left += 1
    else
      scratch[out] = array[right]
      right += 1
    end
    out += 1
  end
  while left < mid
    scratch[out] = array[left]
    left += 1
    out += 1
  end
  while right < finish
    scratch[out] = array[right]
    right += 1
    out += 1
  end
  (start...finish).each { |i| array[i] = scratch[i] }
end

def merge_sort(array, scratch, start, finish)
  return if finish - start < 2
  mid = start + (finish - start) / 2
  merge_sort(array, scratch, start, mid)
  merge_sort(array, scratch, mid, finish)
  merge(array, scratch, start, mid, finish)
end

def sort(array)
  scratch = Array.new(array.length)
  merge_sort(array, scratch, 0, array.length)
  array
end

array = [0, 39, 21, 62, 91, 77, 14, 23,
  90, 69, 51, 81, 68, 83, 32, 56]
sort(array)
p array
